onnxruntime/core/providers/migraphx/gpu_data_transfer.cc

```diff
@@ -57,7 +57,7 @@ common::Status GPUDataTransfer::CopyTensorAsync(const Tensor& src, Tensor& dst,
       HIP_CALL_THROW(hipMemcpyAsync(dst_data, src_data, bytes, hipMemcpyDeviceToDevice, static_cast<hipStream_t>(stream.GetHandle())));
     } else {
       // copy from other CPU memory to GPU, this is blocking
-      HIP_CALL_THROW(hipMemcpy(dst_data, src_data, bytes, hipMemcpyHostToDevice));
+      HIP_CALL_THROW(hipMemcpyWithStream(dst_data, src_data, bytes, hipMemcpyHostToDevice, static_cast<hipStream_t>(stream.GetHandle())));
```

Comment:
why not use hipMemcpyAsync since this function is called CopyTensorAsync?

Reply:
I saw ROCM ep has a similar class:

https://github.com/microsoft/onnxruntime/blob/3ae7c3c0a652b85ac503dfcacaba4ea44c281699/onnxruntime/core/providers/rocm/gpu_data_transfer.cc#L63C27-L63C41

May add a TODO comment to consolidate them to a shared class.

Reply:
> why not use hipMemcpyAsync since this function is called CopyTensorAsync?

Async copy is only available when the CPU memory is pinned. This case is handled by branches a few lines above.
The `hipMemcpy`/`hipMemcpyWithStream` is a fallback when the CPU memory is not pinned.

Reply:
I saw the [hipMemcpyAsync document](https://rocm.docs.amd.com/projects/HIP/en/latest/doxygen/html/group___memory.html#gad55fa9f5980b711bc93c52820149ba18):
`If host or dest are not pinned, the memory copy will be performed synchronously. For best performance, use hipHostMalloc to allocate host memory that is transferred asynchronously.`

I think the hipMemcpyAsync could work when CPU is not pinned.

Also there are some perf issue with hipMemcpyWithStream: https://github.com/ROCm/clr/issues/78

Reply:
I could make a change later.